src/arithmetic/fields.rs

```diff
@@ -4,106 +4,83 @@
 use core::mem::size_of;
 
 use static_assertions::const_assert;
-use subtle::{Choice, CtOption};
+use subtle::Choice;
+
+#[cfg(not(feature = "std"))]
+use subtle::CtOption;
 
 #[cfg(feature = "std")]
 use super::Group;
 
 #[cfg(feature = "std")]
-use std::{
-    assert,
-    boxed::Box,
-    convert::TryInto,
-    io::{self, Read, Write},
-    marker::PhantomData,
-    vec::Vec,
-};
+use std::{assert, boxed::Box, convert::TryInto, marker::PhantomData, vec::Vec};
 
 const_assert!(size_of::<usize>() >= 4);
 
+/// A trait that exposes additional operations related to calculating square roots of
+/// prime-order finite fields.
+#[cfg(feature = "std")]
+pub trait SqrtRatio: ff::PrimeField {
+    /// The value $(T-1)/2$ such that $2^S \cdot T = p - 1$ with $T$ odd.
+    const T_MINUS1_OVER2: [u64; 4];
+
+    /// Raise this field element to the power [`Self::T_MINUS1_OVER2`].
+    ///
+    /// Field implementations may override this to use an efficient addition chain.
+    fn pow_by_t_minus1_over2(&self) -> Self {
+        ff::Field::pow_vartime(&self, &Self::T_MINUS1_OVER2)
+    }
+
+    /// Computes:
+    ///
+    /// - $(\textsf{true}, \sqrt{\textsf{num}/\textsf{div}})$, if $\textsf{num}$ and
+    ///   $\textsf{div}$ are nonzero and $\textsf{num}/\textsf{div}$ is a square in the
+    ///   field;
+    /// - $(\textsf{true}, 0)$, if $\textsf{num}$ is zero;
+    /// - $(\textsf{false}, 0)$, if $\textsf{num}$ is nonzero and $\textsf{div}$ is zero;
+    /// - $(\textsf{false}, \sqrt{G_S \cdot \textsf{num}/\textsf{div}})$, if
+    ///   $\textsf{num}$ and $\textsf{div}$ are nonzero and $\textsf{num}/\textsf{div}$ is
+    ///   a nonsquare in the field;
+    ///
+    /// where $G_S$ ([`ff::PrimeField::root_of_unity`]) is a generator of the order $2^S$
```

Comment:
Btw, when updating the Hashing to Elliptic Curves draft, it was noticed that the SSWU hash-to-curve algorithm can be simplified if the generator used in `sqrt_ratio` is the same as Z in SSWU. So we might change it to be that value. For now, we should just say that users of this crate should not rely on `root_of_unity` not changing.